pkg/kubelet/dockershim/docker_checkpoint.go

```
@@ -0,0 +1,150 @@
+	// GetCheckpoint retrieves sandbox checkpoint from CheckpointStore.
+	GetCheckpoint(podSandboxID string) (*PodSandboxCheckpoint, error)
+	// RemoveCheckpoint removes sandbox checkpoint form CheckpointStore.
+	// WARNING: RemoveCheckpoint will not return error if checkpoint does not exist.
+	RemoveCheckpoint(podSandboxID string) error
+	// ListCheckpoint returns the list of existing checkpoints.
+	ListCheckpoints() []string
+}
+
+// PersistentCheckpointHandler is an implementation of CheckpointHandler. It persists checkpoint in CheckpointStore
+type PersistentCheckpointHandler struct {
+	store CheckpointStore
+}
+
+func NewPersistentCheckpointHandler() CheckpointHandler {
+	return &PersistentCheckpointHandler{store: &FileStore{path: filepath.Join(DockershimRootDir, sandboxCheckpointDir)}}
+}
+
+func (handler *PersistentCheckpointHandler) CreateCheckpoint(podSandboxID string, checkpoint *PodSandboxCheckpoint) error {
+	checkpoint.CheckSum = calculateChecksum(*checkpoint)
+	blob, err := json.Marshal(checkpoint)
+	if err != nil {
+		return err
+	}
+	return handler.store.Write(podSandboxID, blob)
+}
+
+func (handler *PersistentCheckpointHandler) GetCheckpoint(podSandboxID string) (*PodSandboxCheckpoint, error) {
+	blob, err := handler.store.Read(podSandboxID)
+	if err != nil {
+		return nil, err
+	}
+	var checkpoint PodSandboxCheckpoint
+	err = json.Unmarshal(blob, &checkpoint)
+	if err != nil {
+		return &checkpoint, err
+	}
+	if checkpoint.CheckSum != calculateChecksum(checkpoint) {
+		return &checkpoint, ChecksumError
+	}
+	return &checkpoint, nil
+}
+
+func (handler *PersistentCheckpointHandler) RemoveCheckpoint(podSandboxID string) error {
+	if err := handler.store.Delete(podSandboxID); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	return nil
+}
+
+func (handler *PersistentCheckpointHandler) ListCheckpoints() []string {
+	keys, err := handler.store.List()
+	if err != nil {
+		glog.Errorf("Failed to list checkpoint store: %v", err)
```

Why not return the error and let the caller handle it?